plugin/edge/edge.go

```diff
@@ -112,151 +264,130 @@ func (oe *OptikonEdge) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dn
-			return dns.RcodeServerFailure, errTableParseFailure
-		}
+		// Write the response message.
+		w.WriteMsg(res)
 
-		// Remove the Table entry from the return message.
-		ret.Extra = ret.Extra[1:]
+		return dns.RcodeSuccess, nil
+	}
 
-		// If the list is empty, call the next plugin (proxy).
-		if len(edgeSites) == 0 {
-			return plugin.NextOrFailure(oe.Name(), oe.Next, ctx, w, r)
+	// If there was an upstream error, return a server failure.
+	if upstreamErr != nil {
+		if dnsDebugMode {
+			log.Infoln("upstream proxy generated an error (%v)", upstreamErr)
 		}
+		return dns.RcodeServerFailure, upstreamErr
+	}
 
-		// Compute the distance to the first edge site.
-		closest := edgeSites[0].IP
-		minDist := Distance(oe.lat, oe.lon, edgeSites[0].Lat, edgeSites[0].Lon)
-		for _, edgeSite := range edgeSites {
-			dist := Distance(oe.lat, oe.lon, edgeSite.Lat, edgeSite.Lon)
-			if dist < minDist {
-				minDist = dist
-				closest = edgeSite.IP
-			}
-		}
+	// If the request can't be resolved by anything upstream, or if all the upstreams
+	// are unresponsive, fall through to proxy.
+	if dnsDebugMode {
+		log.Infoln("no healthy upstream proxies. falling through to `proxy` plugin")
+	}
+	return plugin.NextOrFailure(e.Name(), e.Next, ctx, w, r)
+}
 
-		// Write the closest cluster IP as a DNS record.
-		var rr dns.RR
-		switch state.Family() {
-		case 1:
-			rr = new(dns.A)
-			rr.(*dns.A).Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeA, Class: state.QClass()}
-			rr.(*dns.A).A = net.ParseIP(closest).To4()
-		case 2:
-			rr = new(dns.AAAA)
-			rr.(*dns.AAAA).Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeAAAA, Class: state.QClass()}
-			rr.(*dns.AAAA).AAAA = net.ParseIP(closest)
-		}
-		ret.Answer = []dns.RR{rr}
+// Write the given IP address as an Authoritative Answer to the request.
+func writeAuthoritativeResponse(res *dns.Msg, state *request.Request, ip net.IP) {
+
+	// Set the reply to the given request.
+	res.SetReply(state.Req)
+
+	// Make the answer Authoritative and compressed.
+	res.Authoritative, res.Compress = true, true
+
+	// Add the IP address to the Answer field.
+	var rr dns.RR
+	switch state.Family() {
+	case 1:
+		rr = new(dns.A)
+		rr.(*dns.A).Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeA, Class: state.QClass()}
+		rr.(*dns.A).A = ip.To4()
+	case 2:
+		rr = new(dns.AAAA)
+		rr.(*dns.AAAA).Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeAAAA, Class: state.QClass()}
+		rr.(*dns.AAAA).AAAA = ip
+	}
+	res.Answer = []dns.RR{rr}
 
-		// Write the response message.
-		w.WriteMsg(ret)
+	// Write the message.
+	state.W.WriteMsg(res)
+}
 
-		return 0, nil
+// Determines the IP address of the edge site closest to the given Point.
+func findClosestToPoint(edgeSiteSet Set, p Point) net.IP {
+	var closest net.IP
+	var minDist float64
+	first := true
+	for _, val := range edgeSiteSet {
+		edgeSite := val.(Site)
+		dist := p.GreatCircleDistance(edgeSite.GeoCoords)
```

Comment:
With lots of sites, this is expensive. We don't have to take this optimization on right now, but lets leave an issue for it.